Migrate to new AMQP API

esdrasjnr opened the pull request:
What does this implement/fix? Explain your changes.
---------------------------------------------------
This patch adapts the connector code to the new AMQP API definitions. Also improve documentation

Where has this been tested?
---------------------------
**Operating System/Platform:** Ubuntu 18.04.4
**NPM Version:** 6.14.5
**NodeJS Version:** 12.15

esdrasjnr committed:
amqp: Add exchange type to AMQP methods

In order to adapt the new API to connector, this patch adds a parameter
to subscribe and publish AMQP methods. This parameter indicates the type
of exchange to publish/subscribe.

esdrasjnr committed:
data: Remove DeviceStore interface

Due to API changes the DeviceStore is useless, once the data validation
with thing schema is not a connector assignment anymore.

esdrasjnr committed:
publisher: Update data operations to the new API

esdrasjnr committed:
docs: Fix new API events payload

This patch adds missing fields in events documentation. Also fixes the
data events exchange name.

netoax commented on src/network/AMQPConnection.js:
```diff
@@ -18,20 +18,20 @@ class AMQPConnection {
-    await this.channel.bindQueue(queue, topic, key);
+  async onMessage(exchangeName, exchangeType, key, callback, noAck) {
+    await this.channel.assertExchange(exchangeName, exchangeType, { durable: true });
+    const { queue } = await this.channel.assertQueue(`${exchangeName}-messages`, { durable: true });
```
I think we can use this moment to improve this queue name by changing it to something like `connector-event-messages`/`connector-command-messages` such as we are now doing in the `babeltower`.

esdrasjnr replied:
Done!

netoax commented on src/data/DeviceStore.js:
```diff
@@ -9,15 +9,15 @@ class DeviceStore {
 
-  async remove(device) {
-    const index = this.loadDevices.indexOf(device);
+  async remove({ id }) {
```
Why not receive only `id`?

esdrasjnr replied:
I just wanted to maintain the signature to avoid unpredictable errors. Do you think it's better to change this function parameter?

netoax replied:
I'm just afraid we are using a lot of destructuring without need. I'd go with receiving only `id`.

esdrasjnr replied:
I've made some changes in PR. Could you validate it please?

netoax replied:
Ack, just add a commit message to _01528b39c42dcb9510e25bbad4c6d59f16108d31_, explaining the reason it's being removed.

esdrasjnr replied:
Done!

netoax commented on src/interactors/UnregisterDevice.js:
```diff
@@ -7,14 +7,15 @@ class UnregisterDevice {
+    if (!error) {
+      try {
+        await this.cloudConnector.removeDevice(device.id);
+        await this.deviceStore.remove(device);
```
According to what I commented [here](https://github.com/CESARBR/knot-fog-connector/pull/56/commits/a533fee29532c98e879d5dc2c8cddc283ccf685a#diff-725648d7905ef61981bc006936345ad1R12), I think you can pass only the `id`.

netoax commented on src/interactors/RegisterDevice.js:
```diff
@@ -7,24 +7,16 @@ class RegisterDevice {
-    } catch (error) {
-      logger.error(error.message);
-      msgResponse = { id, token: '', error: error.message };
+  async execute({ error, ...device }) {
```
Why is this `error` being received and handled here?

esdrasjnr replied:
You're right, moved this check to the handler!

netoax commented on src/network/MessageHandler.js:
```diff
@@ -61,6 +56,13 @@ class MessageHandler {
           exchangeType: 'direct',
         },
       },
+      'data.published': {
```
I think you should create a variable to hold this value as well.

esdrasjnr replied:
Done!

netoax commented on src/network/MessagePublisher.js:
```diff
@@ -1,5 +1,6 @@
 const exchangeName = 'connOut';
 const exchangeControl = 'control';
+const dataExchange = 'data';
```
I'd suggest you rename the commit title to explicit this change is related to the message publisher because we can have data adaptations on the message handler too.

esdrasjnr replied:
Done!

netoax commented on src/services/DevicesService.js:
```diff
@@ -4,8 +4,8 @@ class DevicesService {
     authDeviceInteractor,
-    listDevicesInteractor,
     updateSchemaInteractor,
+    listDevicesInteractor,
```
Please, keep the order of the parameters similar to the definitions order inside the constructor.

esdrasjnr replied:
It's weird, but it was necessary for code to run correctly. Apparently the code was wrong, check [here](https://github.com/CESARBR/knot-fog-connector/blob/f19ec380818bc2f0b07d623c2ef8857f3f8fe784/src/network/MessageHandlerFactory.js#L32) that there was a mismatch in the parameters order.

What intrigues me is that the code should have thrown some error while executing `updateSchema` interactor. By the way, note that doesn't even exist an `listDevices` interactor anymore...

netoax replied:
That makes sense, @esdrasjnr. The problem here is that there's leftover code the list devices interactor, which is confusing us (please remove it). However, notice when we construct this class we pass `updateSchemaInteractor` as its 5ª parameter and it's correctly attributed locally.

netoax reviewed (changes requested):
Thanks for the PR!
I did some general comments to help you improve it.

lcbm commented on src/network/MessagePublisher.js:
```diff
@@ -9,39 +9,39 @@ class MessagePublisher {
 
   async sendRegisteredDevice(body) {
-    await this.queue.send(exchangeName, 'device.registered', body, { Authorization: this.token });
+    await this.queue.send(exchangeName, 'topic', 'device.registered', body, { Authorization: this.token });
```
Despite not having guidelines for this, I don't like the idea of hard-coded values: whenever things change (and they never stop changing), we'll have repetitive refactoring in many parts of the code which could be avoided if we simply add a variable and update its value. Also, it centralizes some of the architectural blueprint (not sure if this is the best word). 

[This](https://github.com/CESARBR/knot-fog-connector/pull/56/commits/d0167beb50ce8f540b89063db26ec3da9ac495f4#diff-842342ee78795fab62f14840a5dece20) seems great.

esdrasjnr replied:
You're right! But since I've made some changes in PR, could you check if is it still necessary?

lcbm reviewed (commented):
Great patch! Couldn't have done it better, hehe
I just made a comment and I'm also keeping an eye out for the discussions on @netoax's comments. Let me know you thoughts on this :smile:

esdrasjnr committed:
handler: Avoid unexpected breaking errors

This patch adds a code check to avoid process end when the handler tries
to execute an invalid method.

esdrasjnr committed:
handler: Update publish data operation

esdrasjnr committed:
services: device: Remove load interactor

Due to API changes, there is no need to maintain a list of devices in
connector instance.

esdrasjnr committed:
handler: Remove auth and list operations

Due to changes in the new API, the connector doesn't handle these
operations anymore. This patch removes the code associated to them both.

esdrasjnr committed:
handler: Update devices service operations

This patch updates the operations exchange, exchange type and routing
keys, as well as the associated service interactors.

esdrasjnr committed:
publisher: Remove uncessary operations

This patch removes all operations not used in the new API. Also, removes
publisher reference from devices service interactors.

esdrasjnr committed:
amqp: Change queue name for connector events

netoax reviewed (approved):
LGTM!
Waiting for tests.

lcbm commented:
Any updates on tests? :thinking:

netoax reviewed (approved):
LGTM!

Ps: this patch will be integrated after the release.
